rootfs/etc/nginx/lua/balancer.lua

```diff
@@ -0,0 +1,112 @@
+    local cookie_value = ngx.var[var_name]
+
+    if cookie_value == nil then
+        local endpoints_roundrobin = ngx.shared.endpoints_roundrobin
+        local ep_index = endpoints_roundrobin:get(http_host)
+        if ep_index == nil then
+            selected_endpoint = backend.endpoints[1]
+            endpoints_roundrobin:set(http_host, 1, 600)
+        else
+            local new_index = ep_index + 1
+            if backend.endpoints[new_index] == nil then
+                selected_endpoint = backend.endpoints[1]
+                endpoints_roundrobin:set(http_host, 1, 600)
+            else
+                selected_endpoint = backend.endpoints[new_index]
+                endpoints_roundrobin:set(http_host, new_index, 600)
+            end
+        end
+        ngx.header['Set-Cookie'] = var_name .. '=' .. tostring(ep_index)
+    else
+        selected_endpoint = backend.endpoints[tonumber(cookie_value)]
+    end
+else
+    if loadbalancealgorithm == "" or loadbalancealgorithm == "round_robin" then
+        local endpoints_roundrobin = ngx.shared.endpoints_roundrobin
+        local ep_index = endpoints_roundrobin:get(http_host)
+        if ep_index == nil then
+            selected_endpoint = backend.endpoints[1]
+            endpoints_roundrobin:set(http_host, 1, 600)
+        else
+            local new_index = ep_index + 1
+            if backend.endpoints[new_index] == nil then
+                selected_endpoint = backend.endpoints[1]
+                endpoints_roundrobin:set(http_host, 1, 600)
+            else
+                selected_endpoint = backend.endpoints[new_index]
+                endpoints_roundrobin:set(http_host, new_index, 600)
+            end
+        end
+    elseif loadbalancealgorithm == "ip_hash" or loadbalancealgorithm == "least_conn" then
+        local endpoints_leastconn = ngx.shared.endpoints_leastconn
+        local ep_index = endpoints_leastconn:get(remote_addr)
+        if ep_index == nil then
+            local counter = {}
+            for k, v in pairs(backend.endpoints) do
+                counter[k] = 0
+            end
+            local leastconn_rows = endpoints_leastconn:get_keys(0)
+            for _, v in pairs(leastconn_rows) do
+                counter[v] = counter[v] + 1
+            end
+            local free_endpoint_id, _ = compare(counter, function(a,b) return a > b end)
+            selected_endpoint = backend.endpoints[free_endpoint_id]
+            endpoints_leastconn:set(remote_addr, free_endpoint_id, 600)
+        else
+            selected_endpoint = backend.endpoints[ep_index]
+            endpoints_leastconn:set(remote_addr, ep_index, 600)
+        end
+    else
+        ngx.status = 503
+        ngx.exit(ngx.status)
+    end
+end
+
+local max_retries = 20
+if selected_endpoint.maxfails ~= 0 then
+    max_retries = selected_endpoint.maxfails
+end
+
+assert(b.set_current_peer(selected_endpoint.hostname, selected_endpoint.port))
+if (selected_endpoint.failtimeout ~= 0) then
```

Comment:
Can you add some logs like `ngx.log(ngx.DEBUG, using endpoint <host>:<port>)` ?

Reply:
done